scRNA-seq/01-normalizing_scRNA-seq.Rmd

```diff
@@ -92,72 +87,133 @@ plot.
+1) Group cells with other like cells by clustering.  
+2) Compute sum factors for each cell cluster grouping.  
+3) Estimate new counts using linear algebra.  
+4) Normalize by the total reads with counts per million reads (CPMs).  
+  
+Because of differences in sequencing depth, it's pretty common to try to correct
+for this by using total per million reads.
+You have probably seen data reported that was, in part, normalized based the 
+total per million reads for a sample such as counts per million reads (CPMs),
+(or TPMs, FPKMs, and RPKMs, which you can [read about more here](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) on your own).
+
+#### Normalization step 1: Estimate cell clusters:
+
+```{r}
+# Calculate clusters (can adjust min.size depending on clusters sizes you may 
+# expect based on prior knowledge)
+qclust <- scran::quickCluster(sce_raw, min.size = 100)
+
+# Look at the group sizes before we continue:
+summary(qclust)
+```
+
+#### Normalization step 2: Compute sum factors:
+
+Computer factors for each sample using the clusters we just made.
+
+```{r}
+# Calculate the size factors based on the clusters
+sce_raw <- scran::computeSumFactors(sce_raw, clusters = qclust)
+```
+
+#### Normalization step 3 and 4: Calculate CPMs and normalize using sum factors:
+
+For more details on the normalization [read here](https://rdrr.io/bioc/scran/man/computeSumFactors.html)
 
 ```{r Normalize the Data}
-seurat_norm <- NormalizeData(object = seurat_obj, 
-                             normalization.method = "LogNormalize",
-                             scale.factor = 10000)
+# Normalize the data
+sce_norm <- scater::calculateCPM(sce_raw, use_size_factors = TRUE)
 ```
 
-Let's check out density plots so we can see how the data has been changed by 
-normalization. 
-  
+## Explore and compare normalized data to counts
+
+Let's recreate the plot we did before, except now, let's put our scran 
+normalized plotted next to it.
+
 ```{r}
 # Get means of raw and normalized datasets
-all_means <- list("counts" = apply(seurat_norm@raw.data, 1, mean), 
-                  "log_norm" = apply(seurat_norm@data, 1, mean))
+all_means <- list("log_counts" = apply(log(1 + counts(sce_raw)), 1, mean), 
+                  "log_scran" = apply(log(1 + sce_norm), 1, mean))
 
 # Make into a dataframe so ggplot likes it
 all_means <- reshape2::melt(all_means) %>%
-  dplyr::mutate("dataset" = factor(L1, levels = c("counts", "log_norm"))) %>% 
+  dplyr::mutate("dataset" = factor(L1, levels = c("log_counts", "log_scran"))) %>% 
   dplyr::select(-L1)
 
 # Plot the density of the means 
 ggplot(all_means, aes(x = value)) +
   geom_density(fill = "lightblue") +
-  facet_wrap(~dataset, scales = "free")
+  facet_wrap(~dataset)
 ```
 
-The log transformed data doesn't look quite as abnormal as the counts data, 
+The scran normalized data doesn't look quite as 0 heavy as the counts data, 
```

Comment:
I think it would be good to have you explain these plots to me in person @cansav09. It is not entirely clear how this plot speaks to this point (lines 123-124)

> So ideally, we would like to avoid the technical zeroes, but leave the biological 
zeroes alone.

Without having you guide me through it.

Reply:
WIthout changing these to CPMs, the graphs look entirely different from what you saw, so I need to go back to the drawing board now and figure out if what I have is what we want, and if it is, I need to figure out how to make graphs that portray that. And how to explain it better.

Reply:
When you say
> WIthout changing these to CPMs
do you mean using `scater::calculateCPM` specifically?

Reply:
Yes. I don't know if there's something I'm doing wrong or what's going on, but using `scran::normalize` barely changes anything as far as the distributions.